File: runtime/worker/src/protocol.ts
export type Request =
  | { type: "ping" }
  | { type: "launch_app"; command: string; args: string[] }
  | { type: "attach_app"; endpoint: string }
  | { type: "switch_window"; target: WindowTarget }
  | { type: "current_window_title" }
  | { type: "click"; window_id: string; locator: LocatorPayload[] }
  | { type: "screenshot"; window_id: string };

export type Response =
  | { type: "pong" }
  | { type: "app_launched"; window_id: string }
  | { type: "app_attached"; window_id: string }
  | { type: "window_switched"; window_id: string; description: string }
  | { type: "current_window_title"; title: string }
  | { type: "clicked" }
  | { type: "screenshot_taken"; path: string }
  | { type: "error"; message: string };

export type WindowTarget =
  | { type: "title"; value: string }
  | { type: "index"; value: number };

export type LocatorPayload =
  | { type: "explicit"; selector: string }
  | { type: "test_id"; value: string }
  | { type: "role_name"; role: string; name: string }
  | { type: "text"; value: string };

File: src/engine/protocol.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum Request {
    Ping,
    LaunchApp {
        command: String,
        args: Vec<String>,
    },
    AttachApp {
        endpoint: String,
    },
    Click {
        window_id: String,
        locator: Vec<LocatorPayload>,
    },
    Screenshot {
        window_id: String,
    },
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum Response {
    Pong,
    AppLaunched { window_id: String },
    AppAttached { window_id: String },
    Clicked,
    ScreenshotTaken { path: String },
    Error { message: String },
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum LocatorPayload {
    Explicit { selector: String },
    TestId { value: String },
    RoleName { role: String, name: String },
    Text { value: String },
}

impl From<crate::steps::Locator> for LocatorPayload {
    fn from(value: crate::steps::Locator) -> Self {
        match value {
            crate::steps::Locator::Explicit(selector) => Self::Explicit { selector },
            crate::steps::Locator::TestId(value) => Self::TestId { value },
            crate::steps::Locator::RoleName { role, name } => Self::RoleName { role, name },
            crate::steps::Locator::Text(value) => Self::Text { value },
        }
    }
}
